Proposed redis plugin

kworm83 opened the pull request:
This is a proposed plugin to get static lease information from redis for review.  Currently only supports IPv4 as I don't have much v6 equipment to test with.

pmazzini commented:
Thanks for your contribution @kworm83!

* Please sign the commit:

```
git commit --amend --signoff
git push --force-with-lease origin redis-plugin
```

* I am worried about the new dependency introduced by the plugin. It will be a dependency even if the plugin is not used. cc @insomniacslk Thoughts?

pmazzini commented on plugins/redis/plugin.go:
```diff
@@ -0,0 +1,147 @@
+// various global variables
+var log = logger.GetLogger("plugins/redis")
+var pool *redis.Pool
+var LeaseTime time.Duration
```
Does this need to be exported? If not start it with lower case.

kworm83 replied:
Oops...I have changed it to lowercase.

kworm83 commented:
@pmazzini that is a good point about the redigo dependency.  I have to admit I am not familiar enough with the go build/module system to know if there is a way to have a conditional dependency like a #define in c.

insomniacslk commented:
> I am worried about the new dependency introduced by the plugin. It will be a dependency even if the plugin is not used. cc @insomniacslk Thoughts?

As long as it doesn't introduce a CGo dependency I'm not against adding new modules. However, importing it in main go means that it will be compiled and shipped even to who doesn't want to use redis, which is not a core feature, rather a nice-to-have for some users.

This problem will rise again, and we should resolve it for everyone. I'm thinking that we could split the core framework and plugins from the  other plugins, by using another subrepo, e.g. coredhcp/extra or something on those lines.

This means what we would have a default main.go with just core plugins, and the users would have to customize it. Some other projects have a builder that lets people specify the plugins to compile in (and under the hood generates a custom main.go), for example Caddy. Some others like CoreDNS have a script to generate it IIRC

insomniacslk commented:
On an unrelated note we should also document the status of each plugin, and whether they implement v6, v4 or both

kworm83 commented:
Personally I like your idea of a sub-repo and users just editing the main.go to add the "extra" modules they want.

pmazzini commented:
> As long as it doesn't introduce a CGo dependency I'm not against adding new modules.

> Personally I like your idea of a sub-repo

Ok. Let move forward and if plugins continue to grow we can move the non core ones to a different repo.

pmazzini commented:
There are some merge conflicts.

pmazzini commented on plugins/redis/plugin.go:
```diff
@@ -0,0 +1,147 @@
+// various global variables
+var log = logger.GetLogger("plugins/redis")
+var pool *redis.Pool
+var leaseTime time.Duration
```
NIT: Better to have one `var`:

```
var (
  log = 
  pool =
)
```

kworm83 replied:
I will fix this

pmazzini commented on plugins/redis/plugin.go:
```diff
@@ -0,0 +1,147 @@
+
+	// Get all options for a MAC
+	options, err := redis.StringMap(conn.Do("HGETALL", "mac:"+req.ClientHWAddr.String()))
+	if err == redis.ErrNil || options["ipv4"] == "" {
```
When is this `redis.ErrNil` returned?

kworm83 replied:
It would be returned when there are no values found for the key.  So basically that is saying if no values are found or there is not an ipv4 address field.

pmazzini replied:
Ok, I would add a comment to clarify the `redis.ErrNil` error.

pmazzini commented on plugins/redis/plugin.go:
```diff
@@ -0,0 +1,147 @@
+	// Set lease time
+	resp.Options.Update(dhcpv4.OptIPAddressLeaseTime(leaseTime))
+
+	// If request has Relay Agent Info copy it to the reply
```
This is not specific to the redis plugin, I think it should be moved to the main handler:
https://github.com/coredhcp/coredhcp/blob/master/coredhcp.go#L169

kworm83 replied:
I was actually thinking that I should add support for a redis field for lease time and if it exists then update it.   Perhaps there could be a default in the main handler and then the redis plugin over-ride it.  What do you think about that?

pmazzini replied:
I was referring to the Relay Agent Info not the lease time. I wouldn't place the lease time in the main handler. Maybe we can have a lease time plugin and the redis plugin over-ride it.

kworm83 replied:
ahhh...sorry...yes the relay agent info is definitely not specific and to be honest I'm not 100% sure it has to be copied according to the RFCs but from wireshark captures the ISC servers copy it.  For the lease time I think the current way of the plugin having a lease time option set in the config file works fine and I could add it to allow redis to override in that plugin.  Do you think I should add support for more options in the redis plugin or just add them when/if anyone else needed them?

kworm83 replied:
Do you want me to make an attempt at moving the Relay Agent Info stuff into the main handler in this pull request?

Natolumin replied:
I think it should go either into the dhcp library instead (github.com/insomniacslk/dhcp), to be included in `dhcpv4.NewReplyFromRequest`; Or in a separate plugin.

The RFC says "DHCP servers claiming to support the Relay Agent Information option SHALL echo the entire contents of the Relay Agent Information option in all replies", but we don't have to support it, so it's fine to just drop it from this

kworm83 replied:
It seems like adding another plugin would not be the best solution if it should be copied if it exists...that could potentially lead to unexpected behavior for a new user switch from another server and missing that they need to enable a plugin.  The dhcp library NewReplyFromRequest seems to be the most logical place for it.  What do you think @insomniacslk ?

pmazzini replied:
Yes, maybe the library is a better place for it. I have seen some relays dropping the packet if the option in not there.

pmazzini replied:
It definitely should not be added here. We accept a library PR :) Otherwise, I can add it.

kworm83 replied:
I can take a stab at the library PR later today.

kworm83 commented:
On the merge conflicts I needed the updated insomniacslk/dhcp version to support the relay agent options.

pmazzini commented:
> On the merge conflicts I needed the updated insomniacslk/dhcp version to support the relay agent options.

That is fine, can you rebase on top of master?

kworm83 commented:
I fixed the variable declaration nit in redis/plugin.go and cleaned up the format of the log messages.  I also re-did redis error checking to differentiate between a hash key not being found vs not having the ipv4 field.  Also removed the relay options copy from the plugin.

pmazzini commented on plugins/redis/plugin.go:
```diff
@@ -0,0 +1,164 @@
+	if args[0] == "" {
+		return nil, nil, errors.New("Redis server can't be empty")
+	}
+	leaseTime, err = time.ParseDuration(args[1])
```
How about moving the default lease time to a separate plugin?

pmazzini replied:
Other than this, it looks good to me.

kworm83 replied:
I don't have a problem with that.  I will remove the default lease time from the redis plugin and make a new PR for a default lease time plugin.

kworm83 commented:
So do I need to do anything else for it to be merged?

pmazzini commented:
I am letting other people comment on it. I'll merge it tomorrow if there are no comments.
